engine/src/witness/btc/source.rs

```diff
@@ -40,22 +47,88 @@ where
 	) -> (BoxChainStream<'_, Self::Index, Self::Hash, Self::Data>, Self::Client) {
 		(
 			Box::pin(stream::unfold(
-				(self.client.clone(), None, make_periodic_tick(POLL_INTERVAL, true)),
-				|(client, last_block_hash_yielded, mut tick)| async move {
+				(
+					self.client.clone(),
+					Option::<BtcSourceState>::None,
+					make_periodic_tick(POLL_INTERVAL, true),
+				),
+				|(client, mut stream_state, mut tick)| async move {
 					loop {
+						// We don't want to wait for the tick if we have backfilling to do, so we do
+						// it here before awaiting the tick.
+						if let Some(state) = &stream_state {
+							if state.best_known_block_height > state.last_block_yielded_index {
+								tracing::debug!(
+									"Backfilling BTC source from index {} to {}",
+									state.last_block_yielded_index,
+									state.best_known_block_height,
+								);
+								let header = client
+									.header_at_index(
+										state.last_block_yielded_index.saturating_add(1),
+									)
+									.await;
+								return Some((
+									header,
+									(
+										client,
+										Some(BtcSourceState {
+											last_block_yielded_hash: header.hash,
+											last_block_yielded_index: header.index,
+											best_known_block_height: state.best_known_block_height,
+										}),
+										tick,
+									),
+								));
+							}
+						}
+
 						tick.tick().await;
 
 						let best_block_header = client.best_block_header().await;
-						if last_block_hash_yielded != Some(best_block_header.hash) {
+
+						let yield_new_best_header: bool = match &mut stream_state {
+							Some(state)
+								if state.last_block_yielded_index.saturating_add(1) ==
```

Comment:
What purpose does the `state.last_block_yielded_index.saturating_add(1) == best_block_header.height` serve? It seems that the hash check should be enough, no?